Change evaluation of cell average data inside limiter

tulioricci opened the pull request:
- Change cell average evalution from "op.elementwise_integral" to "nodal-to-modal, zero-out polynomials, convert back from modal-to-nodal."
- Added function in "simutil" to avoid recomputing temperature for the limited state. Temperature is kept constant during the limiting operation. Thus, there is no need to go over the expensive temperature evaluation if it doesn't change. Instead, all other DVs and TVs (only for N-S eqs.) are re-evaluated for speed-up.
- Add compiled functions to autoignition example.

Work in progress toward applying the limiter during the state creation (https://github.com/illinois-ceesd/mirgecom/pull/743) which will replace the current implementation. However, the current changes are necessary to speed-up stable calculations with multi-species.

**Questions for the review**:
- [ ] Is the scope and purpose of the PR clear?
  - [ ] The PR should have a description.
  - [ ] The PR should have a guide if needed (e.g., an ordering).
- [ ] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

majosm commented on mirgecom/limiter.py:
```diff
@@ -88,14 +92,30 @@ def bound_preserving_limiter(dcoll: DiscretizationCollection, field,
+                                           else 0 for mode_id in grp.mode_ids()]))
+
+    # map from nodal to modal
+    discr = dcoll.discr_from_dd(DD_VOLUME_ALL)
+    modal_map = dcoll.connection_from_dds(DD_VOLUME_ALL, DD_VOLUME_ALL_MODAL)
+    nodal_map = dcoll.connection_from_dds(DD_VOLUME_ALL_MODAL, DD_VOLUME_ALL)
+    modal_field = modal_map(field)
```
```suggestion
    dd_nodal = dd
    dd_modal = dd_nodal.with_discr_tag(DISCR_TAG_MODAL)

    discr = dcoll.discr_from_dd(dd_nodal)
    modal_map = dcoll.connection_from_dds(dd_nodal, dd_modal)
    nodal_map = dcoll.connection_from_dds(dd_modal, dd_nodal)
    modal_field = modal_map(field)
```

majosm commented on mirgecom/limiter.py:
```diff
@@ -88,14 +92,30 @@ def bound_preserving_limiter(dcoll: DiscretizationCollection, field,
+                          cancel_polynomials(grp),
+                          arg_names=("vec", "filter"),
+                          tagged=(FirstAxisIsElementsTag(),))
+              for grp, vec_i in zip(discr.groups, modal_field))
```
Should be a modal `discr`, right?

majosm commented on examples/autoignition-mpi.py:
```diff
@@ -499,25 +500,75 @@ def my_get_timestep(t, dt, state):
+        limited_cv = apply_limiter(
+            fluid_state.cv, fluid_state.pressure, fluid_state.temperature)
+
+        # since temperature is not changed during the limiting process, it is
+        # possible to avoid recomputing it by NOT calling make_fluid_state.
+        # Thus, evaluate ONLY other DV and TV variables and reassemble the state
+        new_dv = update_dv(limited_cv, fluid_state.temperature)
```
@MTCam will be able to provide more specific guidance on this, but I wonder if it would be better to modify `eos.dependent_vars` to take in an optional `temperature` argument instead of adding a separate `update_dependent_vars` function.

majosm commented on examples/autoignition-mpi.py:
```diff
@@ -499,25 +500,75 @@ def my_get_timestep(t, dt, state):
+        # possible to avoid recomputing it by NOT calling make_fluid_state.
+        # Thus, evaluate ONLY other DV and TV variables and reassemble the state
+        new_dv = update_dv(limited_cv, fluid_state.temperature)
+
+        if viscous_terms_on:
+            new_tv = update_tv(limited_cv, new_dv)
+            fluid_state = update_fluid_state(limited_cv, new_dv, new_tv)
+        else:
+            fluid_state = update_fluid_state(limited_cv, new_dv)
+
+        cv = fluid_state.cv
```
Maybe condense this and the `dv` update into a single compiled function, say `construct_fluid_state_with_temperature`? Or even the limiter step too and call it e.g. `limit_fluid_state`?

majosm commented on examples/autoignition-mpi.py:
```diff
@@ -586,9 +637,20 @@ def my_post_step(step, t, dt, state):
+        # same description as in pre-step
+        limited_cv = limiter(fluid_state.cv, fluid_state.pressure,
+                             fluid_state.temperature)
+        new_dv = update_dependent_vars(limited_cv, fluid_state.temperature, eos)
+        if viscous_terms_on:
+            new_tv = gas_model.transport.transport_vars(limited_cv, new_dv, eos)
+            fluid_state = ViscousFluidState(limited_cv, new_dv, new_tv)
+        else:
+            fluid_state = FluidState(limited_cv, new_dv)
```
Same here (though I guess you would call the non-compiled version).

majosm reviewed (commented):
Added a few suggestions/thoughts.

tulioricci commented:
Thanks for your suggestions @majosm . I implemented all points you raised. I also removed the function from "simutil" and placed it in the driver itself, inside the "limit_fluid_state".

majosm reviewed (approved):
Looks good on my end. Let's wait until @MTCam has a chance to look before merging though, as I'm not super familiar with the details of how the limiter works.

MTCam commented on examples/autoignition-mpi.py:
```diff
@@ -499,25 +499,65 @@ def my_get_timestep(t, dt, state):
+        # since temperature nor pressure are changed during the limiting process,
+        # it is possible to avoid recomputing it by NOT calling make_fluid_state.
+        # Thus, evaluate ONLY other DV and TV variables for the limited state
+        new_dv = MixtureDependentVars(
+            temperature=temperature,
+            pressure=pressure,
+            speed_of_sound=eos.sound_speed(limited_cv, temperature),
+            species_enthalpies=eos.species_enthalpies(limited_cv, temperature))
```
Is there a performance reason not to use this:
```suggestion
        new_dv = fluid_state.dv.replace(
               speed_of_sound=eos.sound_speed(limited_cv, temperature)
               species_enthalpies=eos.species_enthalpies(limited_cv, temperature)
        )
```

tulioricci replied:
Not that I am aware. Changing to your suggestion.

MTCam commented on examples/autoignition-mpi.py:
```diff
@@ -499,25 +499,65 @@ def my_get_timestep(t, dt, state):
+
+        # update transport vars, if necessary
+        if gas_model.transport is not None:
+            new_tv = gas_model.transport.transport_vars(limited_cv, new_dv, eos)
```
what about `fluid_state.tv.replace` here too?

tulioricci replied:
In another PR, I am not recomputing the viscosity every time (https://github.com/illinois-ceesd/mirgecom/pull/772/files#diff-cc316056955cf6f5921e330e4134bbbb83ca53f817f90ab48548b3ad4453c028R128). So I think this way is better with that case in mind.
